File: scripts/reference-solutions/minimum-window-substring.py
from collections import Counter, defaultdict
import sys

lines = sys.stdin.read().splitlines()
s = lines[0] if len(lines) > 0 else ""
t = lines[1] if len(lines) > 1 else ""

need = Counter(t)
window = defaultdict(int)
required = len(need)
formed = 0
left = 0
best_start = 0
best_len = None

for right, char in enumerate(s):
    window[char] += 1
    if char in need and window[char] == need[char]:
        formed += 1

    while formed == required and left <= right:
        current_len = right - left + 1
        if best_len is None or current_len < best_len:
            best_start = left
            best_len = current_len

        left_char = s[left]
        window[left_char] -= 1
        if left_char in need and window[left_char] < need[left_char]:
            formed -= 1
        left += 1

if best_len is None:
    print("EMPTY")
else:
    print(s[best_start : best_start + best_len])

File: scripts/reference-solutions/minimum-window-substring.cpp
#include <array>
#include <iostream>
#include <string>

using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  string s;
  string t;
  getline(cin, s);
  getline(cin, t);

  array<int, 256> need{};
  array<int, 256> window{};
  int required = 0;
  for (unsigned char c : t) {
    if (need[c] == 0) {
      ++required;
    }
    ++need[c];
  }

  int formed = 0;
  int left = 0;
  int bestStart = 0;
  int bestLength = static_cast<int>(s.size()) + 1;

  for (int right = 0; right < static_cast<int>(s.size()); ++right) {
    unsigned char c = static_cast<unsigned char>(s[right]);
    ++window[c];
    if (need[c] > 0 && window[c] == need[c]) {
      ++formed;
    }

    while (formed == required && left <= right) {
      int currentLength = right - left + 1;
      if (currentLength < bestLength) {
        bestLength = currentLength;
        bestStart = left;
      }

      unsigned char leftChar = static_cast<unsigned char>(s[left]);
      --window[leftChar];
      if (need[leftChar] > 0 && window[leftChar] < need[leftChar]) {
        --formed;
      }
      ++left;
    }
  }

  if (bestLength == static_cast<int>(s.size()) + 1) {
    cout << "EMPTY\n";
  } else {
    cout << s.substr(bestStart, bestLength) << '\n';
  }

  return 0;
}
